package/MDAnalysis/lib/nsgrid.pyx

```diff
@@ -934,70 +527,69 @@ cdef class FastNS(object):
            :meth:`~NSResults.get_distances`, :meth:`~NSResults.get_pairs`, and
            :meth:`~NSResults.get_pair_distances`.
         """
-
-        cdef ns_int i, j, size_search
-        cdef ns_int d, m
-        cdef ns_int current_beadid, bid
-        cdef ns_int cellindex, cellindex_probe
-        cdef ns_int xi, yi, zi
+        cdef int cx, cy, cz, ox, oy, oz
+        cdef int ci, cj, i, j, nj, size_search
+        cdef int cellindex, cellindex_probe
+        cdef int xi, yi, zi
 
         cdef NSResults results
-        cdef dreal d2
-        cdef rvec probe
+        cdef double d2
 
-        cdef dreal cutoff2 = self.cutoff * self.cutoff
-        cdef ns_int npairs = 0
-        cdef bint check
+        cdef double cutoff2 = self.cutoff * self.cutoff
 
         size_search = self.coords.shape[0]
+        results = NSResults()
 
-        results = NSResults(self.cutoff, self.coords, self.coords)
+        # loop over 13 neighbouring pairs
+        cdef int[:, ::1] route
+        route = np.array([[1, 0, 0], [1, 1, 0], [0, 1, 0], [-1, 1, 0],
```

Comment:
As mentioned above for `search()`: It might be nicer if (relevant) neighbor cells are computed in an extra function.